bindings/python/pgvector_rs/utils/serializer.py

```diff
@@ -0,0 +1,39 @@
+from typing import Optional
+import numpy as np
+from struct import pack, unpack
+from . import decorators
+
+
+@decorators.ignore_none
+@decorators.ignore_ndarray  # could be ndarray if already cast by lower-level driver
+def from_db_str(value):
+    return np.array(value[1:-1].split(","), dtype=np.float32)
+
+
+@decorators.ignore_none
+@decorators.ignore_ndarray  # could be ndarray if already cast by lower-level driver
+def from_db_binary(value):
+    (dim, unused) = unpack(">HH", value[:4])
+    return np.frombuffer(value, dtype=">f", count=dim, offset=4).astype(
+        dtype=np.float32
+    )
+
+
+@decorators.ignore_none
+@decorators.validate_builtin_list
+@decorators.validate_ndarray
+def to_db_str(value: np.ndarray, dim: Optional[int]=None):
+    if dim is not None and value.shape[0] != dim:
+        raise ValueError("expected %d dimensions, not %d" % (dim, len(value)))
+
+    return "[" + ",".join([str(float(v)) for v in value]) + "]"
+
+
+@decorators.ignore_none
+@decorators.validate_builtin_list
+@decorators.validate_ndarray
+def to_db_binary(value: np.ndarray, dim=None):
+    if dim is not None and value.shape[0] != dim:
+        raise ValueError("expected %d dimensions, not %d" % (dim, len(value)))
+
+    return pack(">HH", value.shape[0], 0) + value.astype(">f").tobytes()
```

Comment:
What does this "0" value mean? Padding?

Reply:
This is copied from pgvector-python, and it is unused due to the lack of binary I/O as we discussed in Discord.

```c
typedef struct Vector
{
    int32        vl_len_;        /* varlena header (do not touch directly!) */
    int16        dim;            /* number of dimensions */
    int16        unused;
    float        x[FLEXIBLE_ARRAY_MEMBER];
}            Vector;
```

The 0 refers to the `int16 unused` in pgvector's struct of vector, which is used for memory alignment I guess.